room/compiler/src/main/kotlin/androidx/room/solver/types/CustomTypeConverterWrapper.kt

```diff
@@ -45,13 +50,75 @@ class CustomTypeConverterWrapper(val custom: CustomTypeConverter) :
                         outputVarName, custom.typeName,
                         custom.methodName, inputVarName)
             } else {
-                addStatement("$L = $N.$L($L)",
-                        outputVarName, typeConverter(scope),
+                if (custom.enclosingClassFactory != null) {
+                    addStatement("$L = $N().$L($L)",
+                        outputVarName, typeConverterFactory(scope, custom.enclosingClassFactory),
                         custom.methodName, inputVarName)
+                } else {
+                    addStatement(
+                        "$L = $N.$L($L)",
+                        outputVarName, typeConverter(scope),
+                        custom.methodName, inputVarName
+                    )
+                }
             }
         }
     }
 
+    fun typeConverterFactory(scope: CodeGenScope, enclosingClassFactory: XType): MethodSpec {
+        val baseName = (custom.typeName as ClassName).simpleName().decapitalize(Locale.US)
+
+        val converterField = scope.writer.getOrCreateField(object : ClassWriter.SharedFieldSpec(
+            baseName, custom.typeName) {
+            override fun getUniqueKey(): String {
+                return "converter_${custom.typeName}"
+            }
+
+            override fun prepare(writer: ClassWriter, builder: FieldSpec.Builder) {
+                builder.addModifiers(Modifier.PRIVATE)
+            }
+        })
+
+        return scope.writer.getOrCreateMethod(object : ClassWriter.SharedMethodSpec(
+            baseName) {
+            override fun getUniqueKey(): String {
+                return "converterMethod_${custom.typeName}"
+            }
+
+            override fun prepare(
+                methodName: String,
+                writer: ClassWriter,
+                builder: MethodSpec.Builder
+            ) {
+                builder.apply {
+                    addModifiers(Modifier.PRIVATE)
```

Comment:
add a `Modifier.SYNCHRONIZED` since converter instantiation can race between async queries trying to use it.